### DISCUSSION TITLE
Deleting a cookie by name

### DISCUSSION THREAD

**Discussion by nortxort** [aio-libs/aiohttp#8548] [Q&A] [answered]
So i want to delete a cookie by name. I got no problem accessing the cookies, i just can't figure out how to delete by name.
I am trying to use [clear(predicate)](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/cookiejar.py#L119) i just don't understand how this is to be used. So i ask if someone can show me an example of the correct way of doing this.

The [documentation mentions](https://docs.aiohttp.org/en/stable/abc.html#aiohttp.abc.AbstractCookieJar.clear):

> Removes all cookies from the jar if the predicate is None. Otherwise remove only those [Morsel](https://docs.python.org/3/library/http.cookies.html#http.cookies.Morsel) that predicate(morsel) returns True.

What i find a bit confusing is the 

> Otherwise remove only those..

This to me sounds like it is removing more than one?

Then in the explanation as to what the predicate parameter is, it states 

> predicate – callable that gets [Morsel](https://docs.python.org/3/library/http.cookies.html#http.cookies.Morsel) as a parameter and returns True if this [Morsel](https://docs.python.org/3/library/http.cookies.html#http.cookies.Morsel) must be deleted from the jar.

There it sounds like only one Morsel will be removed.
Can someone show me a code example of how this is to be used?

**Comment by Dreamsorcerer** [comment, marked as answer]
> This to me sounds like it is removing more than one?

Well, yes, the predicate is used to filter which ones to delete.

> There it sounds like only one Morsel will be removed.

The predicate is going to be called once for each Morsel.


This is a very standard approach seen across Python (e.g. the builtin filter() function).

**Reply by Dreamsorcerer** [reply]
It appears Morsel.key is the name: https://docs.python.org/3/library/http.cookies.html#http.cookies.Morsel.key

So, `.clear(lambda m: m.key == "some_name")` is presumably what you want.

**Reply by nortxort** [reply]
@Dreamsorcerer Thank you, for the quick reply.

After looking at the code some more, i realized that i should not be the one providing the Morsel. But rather a function that had all the logic in it to decide what cookie to delete. It took me to long, and i feel rather ashamed.

Also, thanks for the code example. I don't know much about lambda, so i guess ill have to read up on that.
But would your code not delete all cookies, with the key "some_name" no matter what domain they came from?

**Reply by Dreamsorcerer** [reply]
Yes, see the other attributes if you want to restrict it to something else.

A lambda is just an anonymous function, i.e. it is the same as:
```
def foo(m):
    return m.key == "some_name"
```
except that it is never assigned a name.

**Comment by nortxort** [comment]
So i ended up making [this function](https://github.com/nortxort/web/blob/dev/web/session.py#L161) I don't know if it is the most pythonic way of doing it, but it works rather good. I also made a small [test file](https://github.com/nortxort/web/blob/dev/tests/cookies.py) showing a few examples.

Maybe someone else looking for a way of doing this, can use this to get on the right track, and not end up spending as long as i did.

**Reply by Dreamsorcerer** [reply]
Yep, only suggestion is that because the comparison evaluates to True/False, you don't need an if statement (i.e. one liner: `return key == x and domain == y`).
